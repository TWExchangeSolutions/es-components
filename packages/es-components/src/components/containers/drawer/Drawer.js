import React, { Children } from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider, withTheme } from 'styled-components';
import { noop } from 'lodash';
import uncontrollable from 'uncontrollable';
import viaTheme from 'es-components-via-theme';

import DrawerPanel from './DrawerPanel';

function drawerPanelPropType(props, propName, componentName) {
  const value = props[propName];
  const errMsg = `${componentName} ${propName} contains an element that is not a Drawer.Panel.`;

  if (Array.isArray(value) && value.some(({ type }) => type !== DrawerPanel)) {
    return new Error(errMsg);
  }
  if (!Array.isArray(value) && value.type !== DrawerPanel) {
    return new Error(errMsg);
  }
  return null;
}

const StyledDrawer = styled.div`
  background-color: ${props => props.theme.colors.white};
  border-top: 1px solid ${props => props.theme.colors.grayLight};
  box-shadow: 0 1px 1px ${props => props.theme.colors.boxShadowLight};
  margin-bottom: 25px;
`;

export const Drawer = props => {
  const {
    activeKeys,
    children,
    className,
    closedIconName,
    isAccordion,
    onActiveKeysChanged,
    openedIconName,
    theme
  } = props;

  const onItemClick = key => {
    let nextActiveKeys = [...activeKeys];

    if (isAccordion) {
      nextActiveKeys = nextActiveKeys[0] === key ? [] : [key];
    } else {
      const index = nextActiveKeys.indexOf(key);
      const isOpen = index > -1;

      if (isOpen) {
        nextActiveKeys.splice(index, 1);
      } else {
        nextActiveKeys.push(key);
      }
    }

    onActiveKeysChanged(nextActiveKeys);
  };

  const getPanels = () =>
    Children.map(children, (child, index) => {
      // If there is no key provided, use the panel order as default key
      const key = child.key || String(index + 1);
      const { title, titleAside } = child.props;
      const noPadding = child.props.noPadding || false;

      let isOpen = false;
      if (isAccordion) {
        isOpen = activeKeys[0] === key;
      } else {
        isOpen = activeKeys.indexOf(key) > -1;
      }

      const childProps = {
        key,
        title,
        titleAside,
        noPadding,
        isOpen,
        children: child.props.children,
        onItemClick: () => onItemClick(key),
        closedIconName,
        openedIconName
      };

      return React.cloneElement(child, childProps);
    });

  return (
    <ThemeProvider theme={theme}>
      <StyledDrawer className={className}>{getPanels()}</StyledDrawer>
    </ThemeProvider>
  );
};

Drawer.propTypes = {
  /** Set which panels are open */
  activeKeys: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]),
  /** Should only contain one or more Drawer.Panel elements */
  children: drawerPanelPropType,
  /** Add additional CSS classes to the root drawer element */
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  /** Override the default plus icon with another OE icon name */
  closedIconName: PropTypes.string,
  /** Used in uncontrolled mode to set initial drawer state */
  defaultActiveKeys: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]),
  /** Only allows one DrawerPanel to be open at a time */
  isAccordion: PropTypes.bool,
  /** Function called when changing active keys */
  onActiveKeysChanged: PropTypes.func,
  /** Override the default minus icon with another OE icon name */
  openedIconName: PropTypes.string,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object
};

Drawer.defaultProps = {
  activeKeys: [],
  isAccordion: false,
  closedIconName: 'plus',
  onActiveKeysChanged: noop,
  openedIconName: 'minus',
  theme: viaTheme
};

const UncontrolledDrawer = uncontrollable(Drawer, {
  activeKeys: 'onActiveKeysChanged'
});

UncontrolledDrawer.Panel = DrawerPanel;

export default withTheme(UncontrolledDrawer);
