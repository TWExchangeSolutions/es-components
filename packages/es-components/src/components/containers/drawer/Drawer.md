Use `Drawer` and `Drawer.Panel` components to wrap content in collapsable panels. This component can be used in a controlled or uncontrolled way. When uncontrolled, you can specify `defaultActiveKeys` to set initial state. When controlled, you must supply both `activeKeys` and an `onActiveKeysChanged` handler.

### Basic Example (Uncontrolled)
```
<Drawer defaultActiveKeys={"1"}>
  <Drawer.Panel title="Panel 1">
    <p>Pretty much any content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 2">
    <p>More content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </Drawer.Panel>
</Drawer>
```


### Accordion (Uncontrolled)
Add the `isAccordion` property to change the default behavior to an accordion
style, where only one panel can be opened at a time.
```
<Drawer isAccordion>
  <Drawer.Panel title="Accordion Panel 1">
    <p>Pretty much any content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Accordion Panel 2" noPadding>
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </Drawer.Panel>
  <Drawer.Panel title="Accordion Panel 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </Drawer.Panel>
</Drawer>
```


### Advanced Example (Controlled)
Use the `noPadding` property to remove default padding within a panel (useful for lists and tables).
Use the `titleAside` property to display text or other content on the right side of the panel header.
You can customize the `key` property of each Panel; if not specified a default key value will be assigned
matching the Panel's numeric position.
Both the Drawer and Drawer.Panel components will accept additional classNames.
```
class DrawerExample extends React.Component {
  constructor() {
    this.state = { activeKeys: ["second","3"] };
    this.onActiveKeysChanged = this.onActiveKeysChanged.bind(this);
  }

  onActiveKeysChanged(value) {
    this.setState({ activeKeys: value });
  };

  render() {
    return (
      <Drawer activeKeys={this.state.activeKeys} onActiveKeysChanged={this.onActiveKeysChanged} closedIconName="hand-right" openedIconName="hand-down" className="drawer-big">
        <Drawer.Panel title="Panel One" key="first" noPadding titleAside={<em>aside text</em>}>
          <p>The default padding has been removed from this panel.</p>
        </Drawer.Panel>
        <Drawer.Panel title="Panel Two" key="second" noPadding titleAside={<Icon name='tag' />}>
          <ul>
            <li>List Item</li>
            <li>Another Item</li>
          </ul>
        </Drawer.Panel>
        <Drawer.Panel title="Panel Three" className="alt-color">
          <div>
            More content. Anim pariatur cliche reprehenderit, enim eiusmod
            high life accusamus terry richardson ad squid.
          </div>
        </Drawer.Panel>
      </Drawer>
    )
  }
}
<DrawerExample />
```
