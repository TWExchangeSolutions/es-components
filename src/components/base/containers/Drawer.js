import React, { Children, Component } from 'react';
import PropTypes from 'prop-types';
import DrawerPanel from './DrawerPanel';
import { colors } from '../../theme';
import styled from 'styled-components';

function toArray(activeKeys) {
  let currentActiveKeys = activeKeys;
  if (!Array.isArray(currentActiveKeys)) {
    currentActiveKeys = currentActiveKeys ? [currentActiveKeys] : [];
  }
  return currentActiveKeys;
}

function drawerPanelPropType(props, propName, componentName) {
  const value = props[propName];
  const errMsg = `${componentName} ${propName} contains an element that is not a DrawerPanel.`;

  if (Array.isArray(value) && value.some(({ type }) => type !== DrawerPanel)) {
    return new Error(errMsg);
  }
  if (!Array.isArray(value) && value.type !== DrawerPanel) {
    return new Error(errMsg);
  }
  return null;
}

const StyledDrawer = styled.div`
  background-color: ${colors.white};
  border-top: 1px solid ${colors.grayLight};
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.05);
  margin-bottom: 25px;
`;

class Drawer extends Component {
  constructor(props) {
    super(props);

    const currentActiveKeys = this.props.defaultActiveKeys;
    this.state = {
      activeKeys: toArray(currentActiveKeys)
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('activeKeys' in nextProps) {
      this.setState({
        activeKeys: toArray(nextProps.activeKeys)
      });
    }
  }

  onClickItem(key) {
    return () => {
      let activeKeys = [...this.state.activeKeys];

      if (this.props.isAccordion) {
        activeKeys = activeKeys[0] === key ? [] : [key];
      } else {
        const index = activeKeys.indexOf(key);
        const isActive = index > -1;

        if (isActive) {
          activeKeys.splice(index, 1);
        } else {
          activeKeys.push(key);
        }
      }

      this.setState({ activeKeys });
    };
  }

  getPanels() {
    const activeKeys = this.state.activeKeys;

    return Children.map(this.props.children, (child, index) => {
      // If there is no key provided, use the panel order as default key
      const key = child.key || String(index);
      const header = child.props.header;
      const noPadding = child.props.noPadding || false;

      let isActive = false;
      if (this.props.isAccordion) {
        isActive = activeKeys[0] === key;
      } else {
        isActive = activeKeys.indexOf(key) > -1;
      }

      const props = {
        key,
        header,
        noPadding,
        isActive,
        children: child.props.children,
        onItemClick: this.onClickItem(key).bind(this),
        closedIconName: this.props.closedIconName,
        openedIconName: this.props.openedIconName
      };

      return React.cloneElement(child, props);
    });
  }

  render() {
    return (
      <StyledDrawer className={this.props.className}>
        {this.getPanels()}
      </StyledDrawer>
    );
  }
}

Drawer.propTypes = {
  /** Should only contain one or more DrawerPanel elements */
  children: drawerPanelPropType,
  /** Add additional CSS classes to the root drawer element */
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object
  ]),
  /** Override the default plus icon with another OE icon name */
  closedIconName: PropTypes.string,
  /** Specify which panels are opened by default */
  defaultActiveKeys: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]),
  /** Only allows one DrawerPanel to be open at a time */
  isAccordion: PropTypes.bool,
  /** Override the default minus icon with another OE icon name */
  openedIconName: PropTypes.string
};

Drawer.defaultProps = {
  isAccordion: false,
  closedIconName: 'plus',
  openedIconName: 'minus'
};

export default Drawer;
