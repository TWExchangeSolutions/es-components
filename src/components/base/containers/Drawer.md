Use ``Drawer`` and ``DrawerPanel`` components to wrap content in collapsable panels.

### Basic Example:
```
<Drawer>
  <DrawerPanel header="Panel 1">
    <p>Pretty much any content you want in here.</p>
  </DrawerPanel>
  <DrawerPanel header="Panel 2">
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </DrawerPanel>
  <DrawerPanel header="Panel 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </DrawerPanel>
</Drawer>
```

### Advanced Example:
```
<Drawer isAccordion defaultActiveKeys="1" closedIconName="hand-right" openedIconName="hand-down" className="drawer-big">
  <DrawerPanel header="Panel 1" key="1">
    <p>Pretty much any content you want in here.</p>
  </DrawerPanel>
  <DrawerPanel header="Panel 2" key="2" noPadding>
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </DrawerPanel>
  <DrawerPanel header="Panel 3" key="3" className="alt-color">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </DrawerPanel>
</Drawer>
```
