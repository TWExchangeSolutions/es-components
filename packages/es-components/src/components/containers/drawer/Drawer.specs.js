/* eslint-env jest */

import React from 'react';
import { mountWithTheme, renderWithTheme } from '../../../testing';
import Drawer from './Drawer';

jest.mock('../../util/generateAlphaName', () => () => 'abcdef');

const buildDrawer = props => (
  <Drawer className="important" {...props}>
    <Drawer.Panel
      title="collapse 1"
      key="1"
      className="first"
      titleAside="side text"
    >
      first
    </Drawer.Panel>
    <Drawer.Panel title="collapse 2" key="2" className="second" noPadding>
      second
    </Drawer.Panel>
    <Drawer.Panel title="collapse 3" key="3" className="third">
      third
    </Drawer.Panel>
  </Drawer>
);

describe('drawer component', () => {
  describe('drawer', () => {
    it('renders as expected', () => {
      const tree = renderWithTheme(buildDrawer()).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('clicking a closed panel should add panel key to activeKeys', () => {
      const onActiveKeysChanged = jest.fn();
      const drawerInstance = mountWithTheme(
        buildDrawer({ onActiveKeysChanged })
      );
      const firstPanel = drawerInstance.find('.first');
      const panelButton = firstPanel.find('button');

      expect(panelButton.prop('aria-expanded')).toBe(false);

      panelButton.simulate('click');
      expect(onActiveKeysChanged).toBeCalledWith(['1']);
    });

    it('clicking an open panel should remove panel key from activeKeys', () => {
      const onActiveKeysChanged = jest.fn();
      const drawerInstance = mountWithTheme(
        buildDrawer({ activeKeys: ['1'], onActiveKeysChanged })
      );
      const firstPanel = drawerInstance.find('.first');
      const panelButton = firstPanel.find('button');

      expect(panelButton.prop('aria-expanded')).toBe(true);

      panelButton.simulate('click');
      expect(onActiveKeysChanged).toBeCalledWith([]);
    });

    it('should allow a second drawer to be opened', () => {
      const onActiveKeysChanged = jest.fn();
      const drawerInstance = mountWithTheme(
        buildDrawer({ activeKeys: ['1'], onActiveKeysChanged })
      );
      const firstPanel = drawerInstance.find('.first');
      const firstButton = firstPanel.find('button');
      const thirdPanel = drawerInstance.find('.third');
      const thirdButton = thirdPanel.find('button');

      expect(firstButton.prop('aria-expanded')).toBe(true);
      expect(thirdButton.prop('aria-expanded')).toBe(false);

      thirdButton.simulate('click');
      expect(onActiveKeysChanged).toBeCalledWith(['1', '3']);
    });

    it('should allow rendering more than one active drawer', () => {
      const onActiveKeysChanged = jest.fn();
      const drawerInstance = mountWithTheme(
        buildDrawer({ activeKeys: ['1', '3'], onActiveKeysChanged })
      );
      const firstPanel = drawerInstance.find('.first');
      const firstButton = firstPanel.find('button');
      const thirdPanel = drawerInstance.find('.third');
      const thirdButton = thirdPanel.find('button');

      expect(firstButton.prop('aria-expanded')).toBe(true);
      expect(thirdButton.prop('aria-expanded')).toBe(true);
    });
  });

  describe('accordion', () => {
    const isAccordion = true;

    it('renders as expected', () => {
      const tree = renderWithTheme(buildDrawer({ isAccordion })).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should only allow one drawer to be opened at a time', () => {
      const onActiveKeysChanged = jest.fn();
      const accordionInstance = mountWithTheme(
        buildDrawer({ isAccordion, onActiveKeysChanged, activeKeys: ['1'] })
      );
      const firstPanel = accordionInstance.find('.first');
      const firstButton = firstPanel.find('button');
      const thirdPanel = accordionInstance.find('.third');
      const thirdButton = thirdPanel.find('button');

      expect(firstButton.prop('aria-expanded')).toBe(true);
      expect(thirdButton.prop('aria-expanded')).toBe(false);

      thirdButton.simulate('click');
      expect(onActiveKeysChanged).toBeCalledWith(['3']);
    });
  });
});
