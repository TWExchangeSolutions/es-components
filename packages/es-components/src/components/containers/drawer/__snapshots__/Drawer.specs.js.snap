// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawer component accordion renders as expected 1`] = `
<div
  className="important sc-EHOje ixLrxb"
>
  <div
    className="first sc-bwzfXH bBIqkK"
  >
    <div
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat TVtNM"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="sc-bxivhb UZmtw sc-bdVaJa dnSaqI"
            content="65a"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 1
        </span>
        <aside>
          side text
        </aside>
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero sc-ifAKCX HqBXT"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div
        className={undefined}
      >
        first
      </div>
    </div>
  </div>
  <div
    className="second sc-bwzfXH bBIqkK"
  >
    <div
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat TVtNM"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="sc-bxivhb UZmtw sc-bdVaJa dnSaqI"
            content="65a"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 2
        </span>
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero sc-ifAKCX bUSXbP"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div
        className={undefined}
      >
        second
      </div>
    </div>
  </div>
  <div
    className="third sc-bwzfXH bBIqkK"
  >
    <div
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat TVtNM"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="sc-bxivhb UZmtw sc-bdVaJa dnSaqI"
            content="65a"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 3
        </span>
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero sc-ifAKCX HqBXT"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div
        className={undefined}
      >
        third
      </div>
    </div>
  </div>
</div>
`;

exports[`drawer component drawer renders as expected 1`] = `
<div
  className="important sc-EHOje ixLrxb"
>
  <div
    className="first sc-bwzfXH bBIqkK"
  >
    <div
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat TVtNM"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="sc-bxivhb UZmtw sc-bdVaJa dnSaqI"
            content="65a"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 1
        </span>
        <aside>
          side text
        </aside>
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero sc-ifAKCX HqBXT"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div
        className={undefined}
      >
        first
      </div>
    </div>
  </div>
  <div
    className="second sc-bwzfXH bBIqkK"
  >
    <div
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat TVtNM"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="sc-bxivhb UZmtw sc-bdVaJa dnSaqI"
            content="65a"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 2
        </span>
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero sc-ifAKCX bUSXbP"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div
        className={undefined}
      >
        second
      </div>
    </div>
  </div>
  <div
    className="third sc-bwzfXH bBIqkK"
  >
    <div
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat TVtNM"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="sc-bxivhb UZmtw sc-bdVaJa dnSaqI"
            content="65a"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 3
        </span>
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero sc-ifAKCX HqBXT"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div
        className={undefined}
      >
        third
      </div>
    </div>
  </div>
</div>
`;
